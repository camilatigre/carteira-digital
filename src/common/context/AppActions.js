export const makeTransaction = (dispatch, { type, from, to, amountFrom, amountTo }) => {
  dispatch({
    type: 'MAKE_TRANSACTION',
    payload: { type, from, to, amountFrom, amountTo },
  });
};

export const setCoins = (dispatch, { type, buy, sell, amount }) => {
  dispatch({
    type: 'SET_COINS',
    payload: { type, buy, sell, amount },
  });
};

export const updateCoinsAmount = (dispatch, objAmountFrom, objAmountTo) => {
  dispatch({
    type: 'UPDATE_COINS',
    payload: { objAmountFrom, objAmountTo },
  });
};
