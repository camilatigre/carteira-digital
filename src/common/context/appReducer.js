const reducer = (state, action) => {
  switch (action.type) {
    case 'MAKE_TRANSACTION': {
      const { type, from, to, amountFrom, amountTo } = action.payload;
      let transactions = state.transactions;

      transactions = transactions.concat({ type, from, to, amountFrom, amountTo });

      return {
        ...state,
        transactions,
      };
    }

    case 'SET_COINS': {
      let coin = action.payload.type;
      let buy = action.payload.buy;
      let sell = action.payload.sell;
      let amount = action.payload.amount;

      let attCoin = {
        [coin]: {
          buy,
          sell,
          amount,
        },
      };
      return {
        ...state,
        coins: Object.assign(state.coins, attCoin),
      };
    }

    case 'UPDATE_COINS': {
      const { typeFrom, amountFrom } = action.payload.objAmountFrom;
      const { typeTo, amountTo } = action.payload.objAmountTo;

      const coins = state.coins;
      coins[typeFrom]['amount'] = amountFrom <= 0 ? 0 : amountFrom;
      coins[typeTo]['amount'] = amountTo <= 0 ? 0 : amountTo;

      return {
        ...state,
        coins,
      };
    }
    default:
      return state;
  }
};

export default reducer;
