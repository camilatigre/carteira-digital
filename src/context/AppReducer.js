export default (state, action) => {
  console.log(state);
  switch (action.type) {
    case "SAVE_CREDENTIALS":
      return {
        ...state,
        credentials: state.credentials,
      };
    default:
      return state;
  }
};
