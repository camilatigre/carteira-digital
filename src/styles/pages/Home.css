.actionBox {
  width: 100%;
}

.transactions .line {
  margin: 15px 0 5px 0;
}

.MuiButton-outlined.actionSelected {
  background-color: #00ab63;
  color: #fff;
  border-color: #00ab63;
}

.MuiButton-outlined.actionSelected:hover {
  background-color: #00ab63;
  color: #fff;
  border-color: #00ab63;
}

.coinGroup {
  display: flex;
  justify-content: space-between;
  margin: 30px 0 0 0;
}

.coinGroup .select {
  margin: 0 30px 0px 0px;
  width: 100%;
}

.MuiButton-outlined.secondaryButton.lastActionButton {
  margin: 20px 0 0 0;
}
