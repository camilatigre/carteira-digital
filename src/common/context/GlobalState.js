import React, { createContext, useReducer, useContext } from 'react';
import appReducer from './appReducer';
import PropTypes from 'prop-types';

const initialState = {
  transactions: [],
  coins: {
    reais: {
      amount: 100000,
    },
    bitcoins: {
      buy: 64656.45,
      sell: 64656.45,
      amount: 0,
    },
    brita: {
      buy: 5.57,
      sell: 5.57,
      amount: 0,
    },
  },
};

export const GlobalContext = createContext();
export const GlobalDispatch = createContext();
export const GlobalProvider = ({ children }) => {
  const [state, dispatch] = useReducer(appReducer, initialState);

  return (
    <GlobalContext.Provider value={state}>
      <GlobalDispatch.Provider value={dispatch}>{children}</GlobalDispatch.Provider>
    </GlobalContext.Provider>
  );
};

GlobalProvider.propTypes = {
  children: PropTypes.node.isRequired,
};

export const useGlobalState = () => useContext(GlobalContext);
export const useDispatch = () => useContext(GlobalDispatch);
