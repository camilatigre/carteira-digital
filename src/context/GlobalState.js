import React, { createContext, useReducer } from "react";
import AppReducer from "./AppReducer";

const initialState = {
  credenditals: { login: "" },
};

export const GlobalContext = createContext(initialState);
export const GlobalProvider = ({ children }) => {
  const [state, dispatch] = useReducer(AppReducer, initialState);

  function saveCredentials(login) {
    dispatch({
      type: "SAVE_CREDENTIALS",
      payload: login,
    });
  }

  return (
    <GlobalContext.Provider
      value={{
        credenditals: state.credenditals,
        saveCredentials,
      }}
    >
      {children}
    </GlobalContext.Provider>
  );
};
