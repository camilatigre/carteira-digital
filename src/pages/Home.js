import React, { Fragment, useContext } from "react";
import Header from "../common/Header";
import { GlobalContext } from "../context/GlobalState";
import { useHistory } from "react-router-dom";

const Home = () => {
  const { credentials } = useContext(GlobalContext);
  let history = useHistory();

  const reais = 100;
  const isLogged = localStorage.getItem("login") === credentials;

  if (!isLogged) {
    history.push("/401");
  }

  return (
    <Fragment>
      <Header />
      <div id="home">
        <div>Olá, você tem: R$ {reais},00 para comprar Britas e Bitcoins.</div>
      </div>
    </Fragment>
  );
};

export default Home;
