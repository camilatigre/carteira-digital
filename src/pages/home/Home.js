import React, { useEffect } from 'react';
import Header from '../../common/Header';
import { useGlobalState, useDispatch } from '../../common/context/GlobalState';
import { useHistory } from 'react-router-dom';
import '../../styles/common/InternalPages.css';
import { getToday } from '../../utils/dates';
import Transactions from './Transactions';
import Actions from './Actions';
import Prices from './Prices';
import { setCoins } from '../../common/context/AppActions';

const Home = () => {
  let { today, priceDate } = getToday();

  const { coins } = useGlobalState();
  const dispatch = useDispatch();

  useEffect(() => {
    const uriBitcoins = 'https://www.mercadobitcoin.net/api/BTC/ticker/';
    const uriBrita = decodeURI(
      `https://olinda.bcb.gov.br/olinda/servico/PTAX/versao/v1/odata/CotacaoDolarPeriodo(dataInicial=@dataInicial,dataFinalCotacao=@dataFinalCotacao)?%40dataInicial=%27${priceDate}%27&%40dataFinalCotacao=%27${priceDate}%27&%24format=json`,
    );

    fetch(uriBitcoins, {
      method: 'GET',
      headers: [['Content-Type', 'application/json']],
    }).then((res) =>
      res.json().then(({ ticker }) =>
        setCoins(dispatch, {
          type: 'bitcoins',
          buy: parseFloat(ticker.buy).toFixed(2),
          sell: parseFloat(ticker.sell).toFixed(2),
          amount: coins.bitcoins.amount,
        }),
      ),
    );

    fetch(uriBrita, {
      method: 'GET',
      headers: [['Content-Type', 'application/json']],
    }).then((res) =>
      res.json().then(({ value }) => {
        const buyCost = value[0] ? parseFloat(value[0].cotacaoCompra).toFixed(2) : 5.57;
        const sellCost = value[0] ? parseFloat(value[0].cotacaoVenda).toFixed(2) : 5.57;
        setCoins(dispatch, {
          type: 'brita',
          buy: buyCost,
          sell: sellCost,
          amount: coins.brita.amount,
        });
      }),
    );
  }, []);

  let history = useHistory();

  const isLogged = localStorage.getItem('login') !== '';

  if (!isLogged) {
    history.push('/401');
    return;
  }

  return (
    <div className="internalBody">
      <Header />
      <div id="internalPages" className="home">
        <div className="info">
          <div className="container">
            <h3>Olá, {'email@exemplo.com'},</h3>
            <div className="balance">
              <p>seu saldo disponível é:</p>
              <h2>{`R$ ${coins.reais.amount},00`}</h2>
              <h4>Bitcoins: $ {coins.bitcoins.amount}</h4>
              <h4>Britas: $ {coins.brita.amount}</h4>
            </div>
            <Actions />
          </div>
          <Prices today={today} />
        </div>
        <div className="container statement">
          <h3>Extrato de Transações</h3>
          <Transactions />
        </div>
      </div>
    </div>
  );
};

export default Home;
