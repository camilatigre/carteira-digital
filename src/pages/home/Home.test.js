import React from 'react';
import Home from './Home';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import { useGlobalState } from '../../common/context/GlobalState';

jest.mock('../../common/context/GlobalState');

describe('<Home />', () => {
  it('Deveria renderizar a pagina inicial', () => {
    useGlobalState.mockReturnValue({
      coins: {
        reais: {
          amount: 100000,
        },
        bitcoins: {
          buy: 64656.45,
          sell: 64656.45,
          amount: 0,
        },
        brita: {
          buy: 5.57,
          sell: 5.57,
          amount: 0,
        },
      },
    });

    const component = shallow(<Home />);
    expect(toJson(component)).toMatchSnapshot();
  });
});
