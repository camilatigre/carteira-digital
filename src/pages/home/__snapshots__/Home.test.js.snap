// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Home /> Deveria renderizar a pagina inicial 1`] = `
<div
  className="internalBody"
>
  <Header />
  <div
    className="home"
    id="internalPages"
  >
    <div
      className="info"
    >
      <div
        className="container"
      >
        <h3>
          Olá, 
          email@exemplo.com
          ,
        </h3>
        <div
          className="balance"
        >
          <p>
            seu saldo disponível é:
          </p>
          <h2>
            R$ 100000,00
          </h2>
          <h4>
            Bitcoins: $ 
            0
          </h4>
          <h4>
            Britas: $ 
            0
          </h4>
        </div>
        <Actions />
      </div>
      <Prices
        today="27/08/2020"
      />
    </div>
    <div
      className="container statement"
    >
      <h3>
        Extrato de Transações
      </h3>
      <Transactions />
    </div>
  </div>
</div>
`;
